File: programmers/Lv2/영어 끝말잇기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <map>

using namespace std;
map<string,int>m;
bool isAlreadyCalled(string str){
    auto it = m.find(str);
    if(it != m.end()){
        return m[str] > 1;
    }else{
        return false;
    }
}
vector<int> solution(int n, vector<string> words) {
    vector<int> answer;
    int round = 1;
    int number = 2;
    int idx = 0;
    string prev = words[0];
    bool flag = false;
    
    for(int i=1;i<words.size();i++){
        string cur = words[i];
        m[prev]++;
        m[cur]++;
        
        char last = prev[prev.size()-1];
        char first = cur[0];
        
        if(last != first || cur.size() == 1 || isAlreadyCalled(cur)){
            flag = true;
            answer.push_back(number);
            answer.push_back(round);
            break;
        }
        // init
        prev = words[i];
        if(number == n){
            number = 1;
            round++;
        }
        else{
            number++;
        }
    }
    if(flag){
        return answer;
    }
    answer.push_back(0);
    answer.push_back(0);
    return answer;
}

File: programmers/Lv2/영어 끝말잇기.js
function solution(n, words) {
    var answer = [];
    
    const history = [];

    for(let i=0;i<words.length;i++){
        if(i == 0){
            history.push(words[i]);
            continue;
        }
        if(history.includes(words[i])){
            return [(i % n)+1,Math.floor(i / n)+1];
        }
        if(words[i-1].at(words[i-1].length-1) !== words[i].at(0)){
            return [(i % n)+1,Math.floor(i / n)+1];
        }
        history.push(words[i])
    }

    return [0,0];
}
